security/advancedtls/pemfile_provider.go

```diff
@@ -0,0 +1,179 @@
+	if !ok {
+		return nil, fmt.Errorf("failed to call AppendCertsFromPEM")
+	}
+	return trustPool, nil
+}
+
+var logger = grpclog.Component("advancedtls")
+
+// PEMFileProviderOptions contains options to configure a PEMFileProvider.
+// Note that these fields will only take effect in construction time. Once the PEMFileProvider starts, changing fields in PEMFileProviderOptions will do nothing.
+type PEMFileProviderOptions struct {
+	// CertFile is the file path that holds identity certificate whose updates will be captured by a watching goroutine.
+	// Optional. If this is set, KeyFile must also be set.
+	CertFile string
+	// KeyFile is the file path that holds identity private key whose updates will be captured by a watching goroutine.
+	// Optional. If this is set, CertFile must also be set.
+	KeyFile string
+	// TrustFile is the file path that holds trust certificate whose updates will be captured by a watching goroutine.
+	// Optional.
+	TrustFile string
+	// IdentityInterval is the time duration between two credential update checks for identity certs.
+	// Optional. If not set, we will use the default interval.
+	IdentityInterval *time.Duration
+	// RootInterval is the time duration between two credential update checks for root certs.
+	// Optional. If not set, we will use the default interval.
+	RootInterval *time.Duration
+}
+
+// PEMFileProvider implements certprovider.Provider.
+// It provides the most up-to-date identity private key-cert pairs and/or root certificates.
+type PEMFileProvider struct {
+	identityDistributor *certprovider.Distributor
+	rootDistributor     *certprovider.Distributor
+	cancel              context.CancelFunc
+}
+
+// NewPEMFileProvider uses PEMFileProviderOptions to construct a PEMFileProvider.
+func NewPEMFileProvider(o *PEMFileProviderOptions) (*PEMFileProvider, error) {
+	if o.CertFile == "" && o.KeyFile == "" && o.TrustFile == "" {
+		return nil, fmt.Errorf("at least one credential file needs to be specified")
+	}
+	if keySpecified, certSpecified := o.KeyFile != "", o.CertFile != ""; keySpecified != certSpecified {
+		return nil, fmt.Errorf("private key file and identity cert file should be both specified or not specified")
+	}
+	if o.IdentityInterval == nil {
+		o.IdentityInterval = &defaultIdentityInterval
+	}
+	if o.RootInterval == nil {
+		o.RootInterval = &defaultRootInterval
+	}
+	provider := &PEMFileProvider{}
+	if o.CertFile != "" && o.KeyFile != "" {
+		provider.identityDistributor = certprovider.NewDistributor()
+	}
+	if o.TrustFile != "" {
+		provider.rootDistributor = certprovider.NewDistributor()
+	}
+	// A goroutine to pull file changes.
+	identityTicker := time.NewTicker(*o.IdentityInterval)
+	rootTicker := time.NewTicker(*o.RootInterval)
+	ctx, cancel := context.WithCancel(context.Background())
+	// We pass a copy of PEMFileProviderOptions to the goroutine in case users change it after we start reloading.
+	go func(ctx context.Context, o PEMFileProviderOptions) {
```

Comment:
Both the context and the options struct need not be passed to the goroutine.

Reply:
Just to confirm, we don't want users to be able to change interval values after our application starts. This can't happen with the latest code, because in NewPemFileProvider function we are not accepting a pointer of PEMFileProviderOptions, which will force a copy of PEMFileProviderOptions. Is that right?